nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py

```diff
@@ -0,0 +1,665 @@
+                max_num_samples=num_samples[0],
+                seed=data_cfg.get('seed', 1234),
+                context_key=data_cfg.get('context_key', 'text'),
+                label_key=data_cfg.get('label_key', 'answer'),
+                separate_prompt_and_response_with_newline=data_cfg.get(
+                    'separate_prompt_and_response_with_newline', True
+                ),
+                answer_only_loss=self.cfg.get('answer_only_loss', True),
+                truncation_field=data_cfg.get('truncation_field', 'context'),
+                index_mapping_dir=data_cfg.get('index_mapping_dir', None),
+                assistant_prompt=data_cfg.get('assistant_prompt', None),
+            )
+            datasets.append(dataset)
+
+        if is_train:
+            dataset = BlendableDataset(
+                datasets=datasets, weights=data_cfg.concat_sampling_probabilities, size=num_train_samples_after_blend
+            )
+            return dataset
+        else:
+            return datasets
+
+    def _determine_log_key(self, data_config, dataloader_idx, metric_name, mode):
+        # Function that determines whether to log based on the user provided name of the dataset or the dataloader index.
+        base_key = f"{mode}_{metric_name}_" if metric_name is not None else f"{mode}_"
+        # If the user provided names for each validation/test dataset, use those.
+        if hasattr(data_config, "names") and data_config.names is not None:
+            # With only a single validation/test dataset, the name is not a list.
+            if not isinstance(data_config.names, ListConfig):
+                name = data_config.names
+            else:
+                name = data_config.names[dataloader_idx]
+            return base_key + name
+        else:
+            return base_key + f"dataloader{dataloader_idx}"
+
+    def validation_step(self, dataloader_iter, batch_idx, dataloader_idx=0):
+        return self.inference_step(dataloader_iter, batch_idx, 'validation', dataloader_idx)
+
+    def validation_epoch_end(self, outputs):
+        _ = self.inference_epoch_end(outputs, 'validation', self.cfg.data.validation_ds)
+
+    def test_step(self, dataloader_iter, batch_idx, dataloader_idx=0):
+        return self.inference_step(dataloader_iter, batch_idx, 'test', dataloader_idx)
+
+    def test_epoch_end(self, outputs):
+        _ = self.inference_epoch_end(outputs, 'test', self.cfg.data.test_ds)
+
+    def inference_step(self, dataloader_iter, batch_idx, mode, dataloader_idx=0):
+        # Call parent validation step to get the loss.
+        loss = super().validation_step(dataloader_iter, batch_idx)
+        # NOTE: This is a hack to to sidestep inference when the user wants to just monitor the loss only.
+        # TODO (sandeepsub): This should be a better check.
+        return {
+            'loss': loss,
+            'preds': None,
+            'labels': None,
+            'inputs': None,
+        }
```

Comment:
Don't you always return here?

Reply:
Yeah I need to update the comment and leave a TODO to figure out generation in a subsequent PR.